### PR TITLE
(closes #1986) Fixes LFRic OpenMP issues by skipping dep_analysis on PSy coloured loops

### LINKED ISSUE #1986: LFRic OpenMP-generation currently is failing
Some recent change (I am not sure if my OpenMP updates or @hiker converting kernel arguments to PSyIR, or a mix of both) made the LFRic OpenMP compilation fail with:

```
File "/home/sergi/workspace/stfc/PSyclone/src/psyclone/psyir/transformations/parallel_loop_trans.py", line 155, in validate
psyclone.psyir.transformations.transformation_error.TransformationError:
Transformation Error: Dependency analysis failed with the following messages:
Error: The write access to 'call columnwise_op_asm_kernel_code(cmap(colour,cell), nlayers, ncell_2d, div_v_proxy%ncell_3d, div_v_proxy%local_stencil, div_v_cma_matrix, div_v_cma_nrow, div_v_cma_ncol, div_v_cma_bandwidth, div_v_cma_alpha, div_v_cma_beta, div_v_cma_gamma_m, div_v_cma_gamma_p, ndf_aspc1_div_v, cbanded_map_aspc1_div_v, ndf_aspc2_div_v, cbanded_map_aspc2_div_v)
' causes a write-write race condition.
```

This is probably a good thing. The transformation detects that there is actually a race condition (if it weren't for the coloring) which was not detected before.

Conveniently @arporter has a PR that introduces an option to disregard the dependency analysis on the OpenMP validation (#1925). I think Dynamo0p3OMPLoopTrans should pass this option when calling the super class apply (current line 733), after if has validated that the transformation is applied to a colored loop.

### PR DESCRIPTION
**Pull request by sergisiso**

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter This is ready to review. It fixes the current issues with LFRic OpenMP in master.

(Unfortunately it reverses some LFRic loop bounds to Literal['NotInitialized'] I think because the better boundaries were given by some call inside the dependency_analysis call stack)

**Review by arporter — CHANGES_REQUESTED** [review]
Sorry to be a pain but I'd like to understand why changes to the validation have caused a change in the loop bounds displayed in the PSyIR. That doesn't seem to be good/expected behaviour so it would be good to get to the bottom of it.
Apart from that there are just a few minor things to tidy.
Tests and examples are all OK with compilation.
No need to update documentation.
Issue number is not mentioned anywhere.

File: src/psyclone/transformations.py [hunk 1]
```diff
@@ -744,6 +744,7 @@ def apply(self, node, options=None):
 
         :raise TransformationError: if an OMP loop transform would create \
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/transformations.py:745]: Fixed

File: src/psyclone/transformations.py [hunk 2]
```diff
@@ -729,9 +729,9 @@ def __init__(self, omp_schedule="static"):
     def __str__(self):
         return "Add an OpenMP DO directive to a Dynamo 0.3 loop"
 
-    def apply(self, node, options=None):
-        '''Perform Dynamo 0.3 specific loop validity checks then call
-        :py:meth:`OMPLoopTrans.apply`.
+    def validate(self, node, options=None):
+        ''' Perform Dynamo 0.3 specific loop validity checks for the
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/transformations.py:733]: Done

File: src/psyclone/transformations.py [hunk 3]
```diff
@@ -764,7 +769,35 @@ def apply(self, node, options=None):
                 f"Error in {self.name} transformation. The kernel has an "
                 f"argument with INC access. Colouring is required.")
 
-        OMPLoopTrans.apply(self, node, options)
+    def apply(self, node, options=None):
+        ''' Apply Dynamo 0.3 specific OMPLoopTrans.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/transformations.py:773]: Done

File: src/psyclone/transformations.py [hunk 4]
```diff
@@ -764,7 +769,35 @@ def apply(self, node, options=None):
...
+
+        :param node: the Node in the Schedule to check
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/transformations.py:775]: Done

**Comment by sergisiso** [comment]
@arporter This is ready for next review.

> Sorry to be a pain but I'd like to understand why changes to the validation have caused a change in the loop bounds displayed in the PSyIR. That doesn't seem to be good/expected behaviour so it would be good to get to the bottom of it.

It is because calling DependencyTools.can_loop_be_parallelised() or not in LFRic changes the tree.

Ideally it shouldn't change it, but right now when recursing down the tree at some point it calls a
`loop.start_expr` semantic navigation property, which in LFRic (src/psyclone/dynamo0p3.py::7685) if its set to 'not initialized' it automatically replace the bound expression to a newly created reference.

This should be fixed when we get rid of the 'not initialized' bounds completely, but since we are not there yet as a quickfix I believe we changed the semantic navigation to dynamically change the tree when acessed. I think this is partially discussed in #1876

**Comment by arporter** [comment]
> Ideally it shouldn't change it, but right now when recursing down the tree at some point it calls a `loop.start_expr` semantic navigation property, which in LFRic (src/psyclone/dynamo0p3.py::7685) if its set to 'not initialized' it automatically replace the bound expression to a newly created reference.

OK, thanks for explaining - I'm happy now :-)

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Tests and examples are all OK with compilation.
Ref. guide has new errors but they aren't the fault of this PR (and I've taken the liberty of fixing them as it was just missing continuation symbols in a doc string).
Will proceed to merge.

**Pull request merged**